Safari fixes

tkoukkari opened the pull request:
Fixes/changes these issues:
- Automatic Frame search for Safari
  - Own solution to traverse through frames, as switch_to.parent_frame() works differently in Safari than other browsers 
- Window Management for Safari
  - keeps track of open windows, since window_handles returns handles in random order 
- Support for mac COMMAND key
- Removed "not suitable for mac/safari" tags from most tests
- Added Safari to Mac pipeline

Commit:
Added Safari specific solution for COMMAND key

Commit:
Added js_double_click for safari.
Safari also needs to single-click with js

Commit:
Safari proofed table suites

Commit:
safari proof text.robot

Commit:
Added support for window handling in Safari.

Note: Safari returns window handles in random order, i.e. differently
than other browsers. Therefore we must keep track of open safari windows

Commit:
Added own frame search function for Safari

There is a bug in switch_to.parent_frame() in safari and we need to keep
track of traversed child frames and
find the parent ourselves

Commit:
checked rest of the tests for safari compatibility

Commit:
Added Safari to mac pipeline

Replaces firefox in Mac

Commit:
fix dropdown case, need to hover due to resolution

turunenm commented on QWeb/keywords/input_.py:
```diff
@@ -683,10 +685,20 @@ def press_key(locator, key, anchor="1", timeout='0', **kwargs):
         key = input_handler.check_key(key)
-        input_element.send_keys(key)
+
+        if key[0] == '\ue03d' and util.is_safari():  # COMMAND key workaround on safari
```
Is it possible to use other mod keys together with CMD?

For example
`{SHIFT + \ue03d + a}`

If so, then it would be better to check if CMD is in the keys instead of key[0]
`if util.is_safari() and '\ue03d' in key:`

tkoukkari replied:
mainly meant for copy, paste, select all etc. what you can do in a (safari) browser text field. This should support those use cases. If other use cases arise, let's then handle those.

turunenm commented on QWeb/keywords/input_.py:
```diff
@@ -683,10 +685,20 @@ def press_key(locator, key, anchor="1", timeout='0', **kwargs):
+
+        if key[0] == '\ue03d' and util.is_safari():  # COMMAND key workaround on safari
+            javascript.execute_javascript("arguments[0].focus();", input_element)
+            action.key_down(key[0]) \
+                  .send_keys(key[1]) \
+                  .key_up(key[0]).perform()
```
There are at least two cases which will cause trouble

User sends the only the CMD key -> causes index error
User sends CMD + Other mod + key -> some key will not be sent

```
cmd = '\ue03d'
key.remove(cmd)
action.key_down(cmd) \
         .send_keys(key) \
         .key_up(cmd).perform()
```
It might be necessary to handle the case empty key list separately

tkoukkari replied:
cmd alone will not make any sense, but we can just send it as any other key. Empty key list should not be possible, since then the "key" argument is not given. I'll check for length as well in CMD cases, we will support only "normal" cases that make sense in text field context for now (CMD+C, CMD+A etc.).

tkoukkari replied:
New commit should resolve this

turunenm commented on QWeb/internal/frame.py:
```diff
@@ -207,7 +208,63 @@ def search_from_frames(driver=None, current_frame=None):
-        return search_from_frames()
+
+        # pylint: disable=W0102
+        def search_from_frames_safari(driver=None, current_frame=None, parent_tree=[]):
+            keep_frame = kwargs.get('stay_in_current_frame', CONFIG['StayInCurrentFrame'])
+            if keep_frame:
+                return fn(*args, **kwargs)
+            err = None
+            if not driver:
+                driver = browser.get_current_browser()
+                driver.switch_to.default_content()
+            if current_frame is not None:
+                try:
+                    driver.switch_to.frame(current_frame)
+                    logger.debug('switching to childframe {}'.format(str(fn)))
+                except (StaleElementReferenceException, WebDriverException) as e:
+                    logger.warn(e)
+                    driver.switch_to.default_content()
+                    raise e
+            try:
+                web_element = fn(*args, **kwargs)
+            except QWebElementNotFoundError as e:
+                err = e
+                web_element = None
+            if is_valid(web_element):
+                return web_element
+            start = time.time()
+            timeout = CONFIG['FrameTimeout']
+            while time.time() < timeout + start:
+                frames = fc.check_frames(driver)
```
This is duplicate code with `search_from_frames()`, consider refactoring

tkoukkari replied:
It is safari specific function with few differences. Was left there in purpose. We first had only one and did the same for all browsers, but since switch_to.parent_frame() works correctly on all other browsers we decided to keep the old, working one for other browsers and introduce a safari specific solution only for safari.

Since it's recursive and we need to keep track on traversed frames in safari (parent_tree), we think it's easier to remember 6 months from now when there's strictly safari specific funtion.

turunenm commented on QWeb/internal/frame.py:
```diff
@@ -207,7 +208,63 @@ def search_from_frames(driver=None, current_frame=None):
+                    except WebDriverException as e:
+                        driver.switch_to.default_content()
+                        raise e
+                    config.set_config('FrameTimeout', float(timeout + start - time.time()))
+                    logger.trace('Frame timeout: {}'.format(timeout))
+                if err:
+                    raise err
+                return web_element
+            driver.switch_to.default_content()
+            raise QWebTimeoutError('From frame decorator: Unable to locate element in given time')
```
This is duplicate code `search_from_frames()`, consider refactoring

tkoukkari replied:
same comment as for above. It's intentionally safari specific.

turunenm commented on QWeb/internal/frame.py:
```diff
@@ -207,7 +208,63 @@ def search_from_frames(driver=None, current_frame=None):
+                return web_element
+            start = time.time()
+            timeout = CONFIG['FrameTimeout']
+            while time.time() < timeout + start:
```
I don't understand the purpose of this while loop.
At the end of the loop: L261-263
Either an exception is raised or a web_element is returned which should mean that this while-statement acts like if-statement in reality

tkoukkari replied:
Very old code, not much documentation. I wouldn't dare to touch it without exceptionally good reason and lot's of debugging and testing.

The basic idea is to catch and raise specific exceptions to timeout decorators. Certain exceptions make it to re-try finding elements from all frames etc. Timeout guarantees we will not get stuck, i.e jumps out of the loop once timeout expires.

There are also **no** version of keywords where it's ok that webelement is not found.

turunenm reviewed (commented):
Check comments and consider making changes

Commit:
Added checking length of key combinations for Mac
...CDM key. Only normal operations for a text field are supported, i.e.
copy (CMD+C), paste (CMD+V), select all (CMD+A) etc.

Commit:
Added delay to shadow dom test

Commit:
added hidden text to shadow dom case

turunenm reviewed (approved):
Looks good to me
